File: 01-FEA-Spring/Matlab-Code/ReadInput.m
 function [N_NODE,N_ELEM,N_LOAD,N_PRE_DISP,...
       ELEM_NODE,ELEM_STIFF,...
       FORCE_NODE,FORCE_VAL,DISP_NODE,DISP_VAL] = ReadInput(filen,nnode_ele)

% INPUT:
% filen: name of the file with input values
% nnode_ele: number of nodes per element

% OUTPUT: 
% N_NODE, the (integer) number of nodes
% N_ELEM, the (integer) number of elements
% N_LOAD, the (integer) number of nonzero nodal forces, i.e. Pf
% N_PRE_DISP, the (integer) number of nodes with prescribed displacement, i.e. Up
% ELEM_NODE(nnode_ele, N_ELEM), matrix that contains node numbers (integers) for each element
% ELEM_STIFF(N_ELEM), vector that contains stiffness value (real) for each element
% FORCE_NODE(N_LOAD), vector that contains the node numbers (integer) where forces are applied
% FORCE_VAL(N_LOAD), vector that contains the value of the forces (real) corresponding to the numbers in FORCE_NODE
% DISP_NODE(N_PRE_DISP), vector that contains the node numbers (integer) where boundary conditions are imposed
% DISP_VAL(N_PRE_DISP),  vector that contains the value of the forces (real) corresponding to the numbers in DISP_NODE

% Open input file
fid=fopen(filen,'r');
A = fscanf(fid,'%f');

% read first line
N_NODE = A(1);
N_ELEM = A(2);    
N_LOAD = A(3); 
N_PRE_DISP = A(4);

% initialize vectors
ELEM_NODE = zeros(nnode_ele,N_ELEM); 
ELEM_STIFF = zeros(N_ELEM,1);
FORCE_NODE = zeros(N_LOAD,1); 
FORCE_VAL = zeros(N_LOAD,1);
DISP_NODE = zeros(N_PRE_DISP,1); 
DISP_VAL = zeros(N_PRE_DISP,1); 

count = 5;
% get node mumbers in each element and element stiffness
for i=1:N_ELEM
   ELEM_NODE(1,i)=A(count);
   ELEM_NODE(2,i)=A(count+1);
   ELEM_STIFF(i,1)=A(count+2);
   count = count + 3;
end
% get prescribed force values and corresponding node num
for i=1:N_LOAD
   FORCE_NODE(i,1)=A(count);
   FORCE_VAL(i,1)=A(count+1);
   count = count + 2;
end
% get prescribed displacement values and corresponding node num
for i=1:N_PRE_DISP
   DISP_NODE(i,1)=A(count);
   DISP_VAL(i,1)=A(count+1);
   count = count + 2;
end                         

fclose(fid);
end

File: 01-FEA-Spring/C-Code/ReadInput.cpp
#include <MainFile.h>
using namespace std;

void ReadInput(const std::string& filename, int& n_node, int& n_elem, int& n_load, int& n_pre_disp,
	       const int nnode_ele, arma::imat& elem_node, arma::colvec& elem_stiff,
	       arma::icolvec& force_node, arma::colvec& force_val,
	       arma::icolvec& disp_node, arma::colvec& disp_val)
{


  ifstream fout(filename.c_str());
int      count = 0;

if (fout.is_open() ) {

  fout >> n_node >> n_elem >> n_load >> n_pre_disp;

  elem_node = arma::zeros<arma::imat>(nnode_ele,n_elem);
  elem_stiff = arma::zeros<arma::colvec>(n_elem);
  force_node  = arma::zeros<arma::icolvec>(n_load);
  force_val  = arma::zeros<arma::colvec>(n_load);
  disp_node  = arma::zeros<arma::icolvec>(n_pre_disp);
  disp_val  = arma::zeros<arma::colvec>(n_pre_disp);

  for (int i=0;i<n_elem;i++ ) {

    fout >> elem_node(0,i) >> elem_node(1,i) >> elem_stiff(i);

   }
  for (int i=0;i<n_load;i++ ) {

    fout >> force_node(i) >> force_val(i);

   }
  for (int i=0;i<n_pre_disp;i++ ) {

    fout >> disp_node(i) >> disp_val(i);

   }
}

 fout.close();

}
